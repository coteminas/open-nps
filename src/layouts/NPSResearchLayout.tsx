import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import get from 'lodash.get';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import {
  ThemeProvider,
  createMuiTheme,
  ThemeOptions,
  makeStyles,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';

import { ITarget } from '~/model/Target';
import { IResearch } from '~/model/Research';
import { IConfig } from '~/model/Config';
import { IReviewer } from '~/model/Reviewer';
import { AddThemeOptsDefaults } from '~/util/themeOpts';

const styles = (theme) =>
  createStyles({
    root: {
      height: '100vh',
      maxWidth: '750px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      padding: theme.spacing(2),

      '& > h4': {
        marginBottom: theme.spacing(3),
        textAlign: 'justify',
      },
    },
  });

const useStyles = makeStyles(styles);

export interface LayoutProps extends LayoutSharedProps {
  layoutClasses: WithStyles<ReturnType<typeof styles>>['classes'];
}

interface LayoutSharedProps {
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  researchId: string;
  data: {
    reviewer: IReviewer['meta'];
    target: IReviewer['meta'];
  };
}

interface Props extends LayoutSharedProps {
  mui: ThemeOptions;
}

interface GetServerSideProps {
  ctxResearchIdGetter(ctx: GetServerSidePropsContext): string;
  researchExtraData: { concluded: boolean };
}
export const getServerSidePropsFn = ({
  ctxResearchIdGetter,
  researchExtraData,
}: GetServerSideProps) => async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const researchId = ctxResearchIdGetter(ctx);
  const { connectMongo } = await import('../util/mongo');
  const { Target, Research } = await import('../model');

  await connectMongo();

  const research: IResearch = await Research.findOne({
    _id: researchId,
    ...researchExtraData,
  }).populate('reviewer', { _id: 0 });

  const target: ITarget = research
    ? await Target.findById(research.target).populate('configs', { _id: 0 })
    : null;

  if (!research || !target) {
    return { notFound: true };
  }

  const configs: AnyObject = (target.configs as IConfig[]).reduce(
    (a, c) => ({ ...a, [c.key]: c.values }),
    {}
  );

  return {
    props: {
      mui: configs.mui,
      templates: configs.templates,
      themeOpts: AddThemeOptsDefaults(configs.themeOpts),
      data: {
        reviewer: get(research, 'reviewer.meta', {}),
        target: get(target, 'meta', {}),
      },
      researchId,
    },
  };
};

/* istanbul ignore next */
export const removeLocalJSS = (): void => {
  const jssStyles = document.querySelector('#jss-server-side');
  if (jssStyles) {
    jssStyles.parentElement.removeChild(jssStyles);
  }
};

export const withLayout = (
  Component: React.FC<LayoutProps>
): React.FC<Props> => ({ mui, ...props }): React.ReactElement => {
  const classes: LayoutProps['layoutClasses'] = useStyles();

  React.useEffect(removeLocalJSS, []);

  return (
    <ThemeProvider theme={createMuiTheme(mui)}>
      <CssBaseline />
      <Component {...props} layoutClasses={classes} />
    </ThemeProvider>
  );
};

export default withLayout;
