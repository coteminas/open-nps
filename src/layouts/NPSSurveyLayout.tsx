import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import get from 'lodash.get';
import merge from 'lodash.merge';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import {
  ThemeProvider,
  createMuiTheme,
  ThemeOptions,
  makeStyles,
  WithStyles,
  createStyles,
  Theme,
} from '@material-ui/core/styles';

import { ITarget } from '~/model/Target';
import { ISurvey } from '~/model/Survey';
import { IConfig } from '~/model/Config';
import { AddThemeOptsDefaults } from '~/util/themeOpts';
import { AddDefaultTemplates } from '~/util/templates';

const childrenStyles = ({
  SurveyBoxBorderSize,
  SurveyBoxBorderRadius,
}: ThemeOptionsConfigValues) => (theme: Theme) =>
  createStyles({
    root: {
      maxWidth: '750px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      padding: theme.spacing(4),
      border: `${SurveyBoxBorderSize} solid ${theme.palette.primary.main}`,
      borderRadius: SurveyBoxBorderRadius,
    },
    corePhrase: {
      marginBottom: theme.spacing(3),
      textAlign: 'justify',
    },
    commentPhrase: {
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(1),
      textAlign: 'justify',
    },
    brand: {
      width: '100%',
      marginBottom: theme.spacing(3),

      '& > img': {
        width: '100%',
      },
    },
  });

const styles = (theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      height: '100vh',
      padding: theme.spacing(2),
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    },
  });

const useStyles = makeStyles(styles);

export interface LayoutProps extends LayoutSharedProps {
  layoutClasses: WithStyles<ReturnType<typeof childrenStyles>>['classes'];
}

interface LayoutSharedProps {
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  surveyId: string;
  data: {
    reviewer: AnyObject;
    target: ITarget['meta'];
  };
}

interface Props extends LayoutSharedProps {
  mui: ThemeOptions;
}

interface GetServerSideProps {
  ctxSurveyIdGetter(ctx: GetServerSidePropsContext): string;
  surveyExtraData: { concluded: boolean };
}

const mapConfigsFromList = (configs: IConfig[]) =>
  configs.reduce((a, c) => ({ ...a, [c.key]: c.values }), {});

export const getServerSidePropsFn = ({
  ctxSurveyIdGetter,
  surveyExtraData,
}: GetServerSideProps) => async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const surveyId = ctxSurveyIdGetter(ctx);
  const { connectMongo } = await import('../util/mongo');
  const { Target, Survey } = await import('../model');

  await connectMongo();

  const survey: ISurvey = await Survey.findOne({
    _id: surveyId,
    ...surveyExtraData,
  });

  const target: ITarget = survey
    ? await Target.findById(survey.target).populate('configs', { _id: 0 })
    : null;

  if (!survey || !target) {
    return { notFound: true };
  }

  const overrideConfigsList = await survey.getOverrideConfigs();
  const overrideConfigs: AnyObject[] = overrideConfigsList.map(
    mapConfigsFromList
  );

  const configs: AnyObject = [
    mapConfigsFromList(target.configs as IConfig[]),
    ...overrideConfigs,
  ].reduce((a, b) => merge(a, b), {});

  return {
    props: {
      mui: get(configs, 'mui', {}),
      templates: AddDefaultTemplates(configs.templates),
      themeOpts: AddThemeOptsDefaults(configs.themeOpts),
      data: {
        reviewer: JSON.parse(JSON.stringify(survey.reviewer)),
        target: get(target, 'meta', {}),
      },
      surveyId,
    },
  };
};

/* istanbul ignore next */
export const removeLocalJSS = (): void => {
  const jssStyles = document.querySelector('#jss-server-side');
  if (jssStyles) {
    jssStyles.parentElement.removeChild(jssStyles);
  }
};

export const withLayout = (
  Component: React.FC<LayoutProps>
): React.FC<Props> => ({ mui, ...props }): React.ReactElement => {
  const layoutClasses: LayoutProps['layoutClasses'] = makeStyles(
    childrenStyles(props.themeOpts)
  )();
  const classes = useStyles();

  React.useEffect(removeLocalJSS, []);

  return (
    <ThemeProvider theme={createMuiTheme(mui)}>
      <CssBaseline />
      <div className={classes.root}>
        <Component {...props} layoutClasses={layoutClasses} />
      </div>
    </ThemeProvider>
  );
};

export default withLayout;
