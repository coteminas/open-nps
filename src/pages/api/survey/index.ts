import merge from 'lodash.merge';
import { NextApiRequest, NextApiResponse } from 'next';

import { createApiHandler } from '~/util/api';
import Survey, { ISurvey } from '~/model/Survey';
import Target, { ITarget } from '~/model/Target';

export const createNewSurvey = async (
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> => {
  const { reviewer, tags, targetName } = req.body;
  const target: ITarget = await Target.findOne({ name: targetName });

  const survey: ISurvey = await Survey.create({
    target: target._id,
    tags,
    reviewer,
  });

  return res.json(survey);
};

export default createApiHandler({
  POST: createNewSurvey,
});
