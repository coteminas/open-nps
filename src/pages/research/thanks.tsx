import React from 'react';
import Typography from '@material-ui/core/Typography';

import { GetServerSidePropsContext, GetServerSidePropsResult } from 'next';
import { ThemeOptions } from '@material-ui/core/styles';
import { withStyles, WithStyles, createStyles } from '@material-ui/styles';

import Research from '~/model/Research';
import Target, { ITarget } from '~/model/Target';

import { AddThemeOptsDefaults } from '~/util/themeOpts';
import { template } from '~/util/template';
import { IReviewer } from '~/model/Reviewer';

const styles = () =>
  createStyles({
    root: {
      maxWidth: '800px',
      margin: '0 auto',
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    },
  });

interface Props extends WithStyles<typeof styles> {
  mui: ThemeOptions;
  themeOpts: ThemeOptionsConfigValues;
  templates: TemplatesConfigValues;
  researchId: string;
  data: {
    reviewer: IReviewer;
    target: ITarget;
  };
}

export const getServerSideProps = async (
  ctx: GetServerSidePropsContext
): Promise<GetServerSidePropsResult<Props>> => {
  const { connectMongo } = await import('../../util/mongo');
  const { researchId } = ctx.query;

  await connectMongo();

  const research = await Research.findOne({ _id: researchId })
    .populate('reviewer', { _id: 0 })
    .lean();

  const target = await Target.findOne({ _id: research.target })
    .populate('configs', { _id: 0 })
    .lean();

  const { themeOpts, ...configs } = target.configs.reduce(
    (a, c) => ({ ...a, [c.key]: c.values }),
    {}
  );
  return {
    props: {
      ...configs,
      themeOpts: AddThemeOptsDefaults(themeOpts),
      researchId,
      data: {
        reviewer: (research.reviewer as IReviewer).meta || {},
        target: target.meta,
      },
    },
  };
};

export const ThanksPage: React.FC<Props> = ({ classes, templates, data }) => (
  <div className={classes.root}>
    <Typography variant="h2" component="h2">
      {template(templates.ThanksPhrase, data)}
    </Typography>
  </div>
);

export default withStyles(styles)(ThanksPage);
