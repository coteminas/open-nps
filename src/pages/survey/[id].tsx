import React, { useState } from 'react';
import Typography from '@material-ui/core/Typography';
import get from 'lodash.get';

import { useRouter, NextRouter } from 'next/router';
import { GetServerSidePropsContext } from 'next';

import SurveyNotes from '~/components/SurveyNotes';
import SurveyComment from '~/components/SurveyComment';
import SurveySubmit from '~/components/SurveySubmit';

import { renderTemplate } from '~/util/renderTemplate';
import {
  withLayout,
  LayoutProps,
  getServerSidePropsFn,
} from '~/layouts/NPSSurveyLayout';

export const ctxSurveyIdGetter = (ctx: GetServerSidePropsContext): string =>
  ctx.params.id as string;

export const getServerSideProps = getServerSidePropsFn({
  ctxSurveyIdGetter,
  surveyExtraData: { concluded: false },
});

interface SubmitData {
  surveyId: string;
  note: string;
  comment: string;
}

export const createSubmit = (data: SubmitData, router: NextRouter) => async (
  e: React.FormEvent<HTMLFormElement>
): Promise<void> => {
  e.preventDefault();
  const response = await fetch(
    `${window.location.origin}/api/survey/conclude`,
    {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    }
  );

  const { ok } = await response.json();

  if (ok) {
    router.push(`/survey/thanks?surveyId=${data.surveyId}`);
  }
};

export const setValueForFieldInState = (
  state: AnyObject,
  setState: SimpleFn<AnyObject, void>
) => (field: string) => (value: string): void =>
  setState({ ...state, [field]: value });

export const SurveyPage: React.FC<LayoutProps> = ({
  themeOpts,
  templates,
  data,
  surveyId,
  layoutClasses,
}): React.ReactElement => {
  const [state, setState] = useState({ note: null, comment: '' });
  const router = useRouter();
  const onSubmit = createSubmit({ surveyId, ...state }, router);
  const setValueForField = setValueForFieldInState(state, setState);

  return (
    <form className={layoutClasses.root} onSubmit={onSubmit}>
      {themeOpts.SurveyTopBrandImage && (
        <div
          style={{
            maxWidth: get(themeOpts, 'SurveyTopBrandImage.width', 'auto'),
          }}
          className={layoutClasses.brand}
        >
          <img
            alt={themeOpts.SurveyTopBrandImage.alt}
            src={themeOpts.SurveyTopBrandImage.url}
          />
        </div>
      )}
      <Typography data-cy="SurveyPageTypography" variant="h4" component="h4">
        {renderTemplate(templates.CoreQuestionPhrase, data)}
      </Typography>
      <SurveyNotes
        themeOpts={themeOpts}
        setValue={setValueForField('note')}
        selected={state.note}
      />
      <SurveyComment
        value={state.comment}
        setValue={setValueForField('comment')}
        label={templates.SurveyCommentLabel}
        placeholder={templates.SurveyCommentPlaceholder}
      />
      <SurveySubmit themeOpts={themeOpts}>
        {templates.SendButtonMessage}
      </SurveySubmit>
    </form>
  );
};

export default withLayout(SurveyPage);
