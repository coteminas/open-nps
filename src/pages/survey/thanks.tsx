import React from 'react';
import Typography from '@material-ui/core/Typography';

import { GetServerSidePropsContext } from 'next';

import {
  withLayout,
  getServerSidePropsFn,
  LayoutProps,
} from '~/layouts/NPSSurveyLayout';
import { renderTemplate } from '~/util/renderTemplate';

export const ctxSurveyIdGetter = (ctx: GetServerSidePropsContext): string =>
  ctx.query.surveyId as string;

export const getServerSideProps = getServerSidePropsFn({
  ctxSurveyIdGetter,
  surveyExtraData: { concluded: true },
});

export const ThanksPage: React.FC<LayoutProps> = ({
  layoutClasses,
  templates,
  data,
}) => (
  <div className={layoutClasses.root}>
    <Typography data-cy="ThanksPageTypography" variant="h4" component="h4">
      {renderTemplate(templates.ThanksPhrase, data)}
    </Typography>
  </div>
);

export default withLayout(ThanksPage);
