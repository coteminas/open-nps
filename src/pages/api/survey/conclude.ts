import Survey, { ISurvey } from '~/model/Survey';

import { createApiHandler } from '~/util/api';
import { NextApiRequest, NextApiResponse } from 'next';

export const concludeSurvey = async (
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> => {
  const { note, comment, surveyId } = req.body;

  const survey: ISurvey = await Survey.findOne({
    _id: surveyId,
    concluded: false,
  });

  const response = await survey.updateOne({ concluded: true, note, comment });
  return res.json(response);
};

export default createApiHandler({
  PUT: concludeSurvey,
});
