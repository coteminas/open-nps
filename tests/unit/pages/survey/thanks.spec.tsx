import React from 'react';

import { ThanksPage, ctxSurveyIdGetter } from '~/pages/survey/thanks';
import { LayoutProps } from '~/layouts/NPSSurveyLayout';
import { shallow } from 'enzyme';
import { GetServerSidePropsContext } from 'next';

describe('/src/pages/survey/thanks', () => {
  it('should render properly', () => {
    const props = ({
      templates: {
        ThanksPhrase: 'Foo Bar',
        ThanksSubPhrase: 'Fizz Fuzz',
      },
      surveyId: 'fake',
      themeOpts: {
        ThanksTopImage: {
          url: 'aaaa',
          alt: 'bbbb',
        },
      },
      data: {
        b: 1,
      },
      layoutClasses: {
        root: 'root',
      },
    } as unknown) as LayoutProps;

    const wrap = shallow(<ThanksPage {...props} />);
    expect(wrap).toMatchSnapshot();
  });

  it('ctxSurveyIdGetter', () => {
    const ctx = ({
      query: { surveyId: 'foo' },
    } as unknown) as GetServerSidePropsContext;
    expect(ctxSurveyIdGetter(ctx)).toBe(ctx.query.surveyId);
  });
});
