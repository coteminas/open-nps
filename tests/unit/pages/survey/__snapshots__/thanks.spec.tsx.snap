// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/src/pages/survey/thanks should render properly 1`] = `
<div
  className="root"
>
  <WithStyles(ForwardRef(Typography))
    component="h4"
    data-cy="ThanksPageTypography"
    variant="h4"
  >
    Foo Bar
  </WithStyles(ForwardRef(Typography))>
</div>
`;
